design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,346 @@
+xlang_thread_state         | 0x80131520 | Operation can not be done due to the thread state
+xlang_thread_interrupted   | 0x80131519 | Interrupt on thread
+xlang_security             | 0x8013150A | Security related error
+xlang_type_load            | 0x80131522 | Syntax / type error during runtime
+
+Even though xlang only defines these common errors, it will not prevent the use of other errors from
+the HRESULT family. But, it is likely that those errors will not get a natural mapping from
+projections and end up as generic errors.
+
+## How to indicate the result of a function
+
+Functions declared at the xlang ABI (i.e. language projections, PAL) need a way to indicate whether
+an error occurred. This will be done through the return parameter.
+
+```
+xlang_result functionName(param, ...)
+```
+
+Returning one of the error codes will indicate that the function resulted in an error.
+Returning xlang_ok which has a value of 0 will indicate the function completed without errors.
+
+## Alternative to errors
+
+Even though xlang does provide a set of errors to assist with the error translation from one language
+to another in an equivalent form, it is limited. In addition, when errors are translated from one language
+to another, there is a chance that the accuracy of the error is lost due to an exact error mapping is not
+available in the other language. This means the caller of an API may not always have the accurate and
+necessary information to decide what to do with a failure.
+
+Due to this, xlang encourages the use of errors only for diagnosing failures and not for controlling
+code flow. For controlling code flow, xlang encourages the use of the Try pattern in functions.
+The Try pattern can be used to pass back custom domain specific errors as parameters if necessary.
+The parameter type and its possible values are declared in the producing component and would be translated
+across languages as is. This means the consumer can rely on using it to make decisions without being
+concerned of the translation accuracy.
+
+## Diagnosing application failures
+
+Errors coming from components written in other languages can be one of the causes of application
+failures. Just like with any other failure, developers will want to diagnose where and why the
+failure occurred. To assist with this, xlang will provide the ability for a projection to associate
+diagnostic information with the error. The following information can be associated:
+
+- Language specific error (i.e. error code, exception type)
+- Error message
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+Some of this information can be represented by different languages in different ways. For instance,
+in some languages, stack traces can be an array of addresses while in others it can be a list of
+function names. Due to this, xlang will store the information generically as strings and it will
```

Comment:
should we specify conventions for these string representations to enable tooling?

Reply:
Agreed. In Rome we used a lightweight object model for this where it was a vector of Stack Frames and each frame has a few strings that can tailored to the desired purpose (file name, function, etc). Some structure to this would be helpful in converting I would think.

Reply:
As per arch meeting, @bbowman expanded on how Rome does this. Rome will convert the lightweight OM to something Java specific so that the Java developer can see the stack in a way that feels like a native Java exception.

Reply:
Will address in future PR.